Introduce test_network feature

kpandl opened the pull request:
## Motivation

For testing purposes, it would be practical to test the snarkVM consensus version (and proof targets) heights through a `--features test_network` compilation flag in snarkOS. This PR introduces such a flag.

It also updates the snarkVM ref to the version where https://github.com/ProvableHQ/snarkVM/pull/2693 is merged.

## Test Plan

Tested with an added CLI output of consensus height thresholds for testing purposes, and verified it works.

kpandl committed:
introduce test feature

kaimast commented:
This looks fine to me.

There is one thing to consider, though. Should the `test` feature enable `test` for all crates in the workspace? `ledger-service`, `storage-service`, `router`, and `sync` all have such a feature. 
If not, we should leave a comment somewhere that explains what features/behavior it enables and why. Alternatively, we could name the feature something like `test-consenus-heights` instead. 

###  "Big Picture" Comments
* I dug through these feature flags a while ago when trying to debug #3546. It's a little confusing how the `test` and `test_targets` features are used in some, but not all, unit tests (usually by having the feature set in a dev-dependency).

* We should document all feature flags for the top-level crate/workspace. The Rust documentation points to [this README](https://github.com/rust-lang/regex/blob/1.4.2/src/lib.rs#L488-L583) as a good example of how to do it.

vicsn commented:
> We should document all feature flags for the top-level crate/workspace. The Rust documentation points to this README as a good example of how to do it.

Really good idea. Not sure if its overkill to use a tool to visualize the feature trees. This documentation can also document the discrepancy you noted regarding test and test_targets. CC Perhaps @joske wants to take a stab at this because he also has context on the addition of test_targets.

niklaslong commented:
> Not sure if its overkill to use a tool to visualize the feature trees.

You should be able to get quite far with `cargo tree`: https://doc.rust-lang.org/cargo/commands/cargo-tree.html, specifically the `-e` option used for searching through features, perhaps combined with `-i` for a reverse search.

vicsn commented:
> There is one thing to consider, though. Should the test feature enable test for all crates in the workspace? ledger-service, storage-service, router, and sync all have such a feature.

@kpandl want to give this a try, and see if we can still run a local devnet and land a transaction? No rush on this.

kpandl committed:
all test feature propagation

kpandl commented:
Thanks, I've added the `test` feature to `router`, `sync`, and through feature propagation of `bft` to `ledger-service` and `storage-service`. I've also updated the `router` crate to propagate the test feature to the submodule.

Ran a devnet and verified a transaction still lands when using the `test` feature (didn't test anything else).

kaimast reviewed (dismissed):
Looks good!

vicsn commented on Cargo.toml:
```diff
@@ -82,6 +82,12 @@ locktick = [
   "snarkos-node-tcp/locktick"
 ]
+test = [
+  "snarkos-node-bft/test",
```
Since we're importing snarkVM in this file, shouldn't we be specifying `"snarkvm/test"` here as well? If only for consistency. (Don't ask me why snarkVM is imported in two placed by snarkOS)

kpandl replied:
Addressed here: https://github.com/ProvableHQ/snarkOS/pull/3572/commits/e66c42ee4fcf34e2545503c3295fd143bda56086
I've added `"snarkvm/test"` to the relevant crates that directly depend on snarkVM (specifically `node/bft/Cargo.toml` and `node/router/Cargo.toml`). I couldn't add it to the root Cargo.toml where I got this:

```
error: feature `test` includes `snarkvm/test`, but `snarkvm` is not a dependency
  --> Cargo.toml:85:8
   |
85 |   test = [
   |  ________^
86 | |   "snarkos-node-bft/test",
...  |
90 | |   "snarkvm/test"
91 | | ]
   | |_^
   |
error: failed to parse manifest at `/Users/kp/dev/snarkOS/Cargo.toml`

```

I've kept changes focused on the files already in the PR scope, though there are many more crates where snarkVM is added as a dependency.

kpandl committed:
add snarkvm/test

kpandl commented:
I've added `"snarkvm/test"` to the relevant crates that directly depend on snarkVM (specifically `node/bft/Cargo.toml` and `node/router/Cargo.toml`). I couldn't add it to the root Cargo.toml where I got this:

```
error: feature `test` includes `snarkvm/test`, but `snarkvm` is not a dependency
  --> Cargo.toml:85:8
   |
85 |   test = [
   |  ________^
86 | |   "snarkos-node-bft/test",
...  |
90 | |   "snarkvm/test"
91 | | ]
   | |_^
   |
error: failed to parse manifest at `/Users/kp/dev/snarkOS/Cargo.toml`

```

I've kept changes focused on the files already in the PR scope, though there are many more crates where snarkVM is added as a dependency.

kaimast commented:
Seems like adding the snarkVM test feature broke something.

Does the test feature get enabled in unit tests? If it isn't, these changes shouldn't affect the tests, but it seems like they do.

vicsn commented:
> Seems like adding the snarkVM test feature broke something.
> Does the test feature get enabled in unit tests? If it isn't, these changes shouldn't affect the tests, but it seems like they do.

Indeed, it seems like for some subcrates, `features = [ "test" ]` is enabled in a `dev-dependencies` subsection. Note for example that router had an empty `test` feature which is now suddenly enabled with a lot more junk in there: https://github.com/ProvableHQ/snarkOS/pull/3572/files#diff-f5e82e097aa5005ad9c92fe29abb9de43cddfbf974a5dcafcdad98e4cbb7b715R20

I see a couple of directions to achieve the original goal: ensure we can run a snarkOS network where novel features are enabled on a low block height.
1. We introduce a new `test_consensus_versions` feature in snarkVM and snarkOS, which will use the test consensus version heights.
2. We introduce `test_targets` and `test_consensus_versions` in snarkVM, but use a single feature in snarkOS called `test_network` which will enable both in snarkVM.
3. We consolidate `test_targets` and `test_consensus_versions` into a single feature called `test`, but this requires untangling the morass of `git grep -C2 test | grep Cargo.toml`.

I vote for starting with option 2., and we can create an issue to clean up 3.

kpandl committed:
Revert "add snarkvm/test"

This reverts commit e66c42ee4fcf34e2545503c3295fd143bda56086.

kpandl committed:
Revert "all test feature propagation"

This reverts commit 59eef1575f611cc9482f23a4fabf2ecc00bc325c.

kpandl committed:
Revert "introduce test feature"

This reverts commit 49e4023a96515845e1984a1e3c55bb36ae2981d9.

kpandl committed:
test_network feature

kpandl committed:
Merge branch 'staging' into test-feature

Signed-off-by: Konstantin Pandl <konstantinpandl@yahoo.de>

kpandl commented:
Thanks, addressed option 2 now. Tested it along with https://github.com/ProvableHQ/snarkVM/pull/2693 on a local devnet.
Note that CI will fail because it needs this snarkVM: https://github.com/ProvableHQ/snarkVM/pull/2693

I'd suggest I temporarily change the snarkVM ref here so that we can let CI run and see if it passes.

kpandl committed:
temporary snarkVM ref change for CI

vicsn commented on cli/src/commands/start.rs:
```diff
@@ -523,8 +523,8 @@ impl Start {
         // allowed in --dev mode.
-        if cfg!(feature = "test_targets") && self.dev.is_none() {
-            bail!("The 'test_targets' feature is enabled, but the '--dev' flag is not set");
+        if cfg!(feature = "test_network") && self.dev.is_none() {
```
Can we also add a guard that we're not running mainnet? Because the feature only influences canary/testnet traits (we can question that decision later if it creates trouble).

vicsn commented on Cargo.toml:
```diff
@@ -82,7 +82,10 @@ locktick = [
-test_targets = [ "snarkos-cli/test_targets" ]
+test_network = [
+  "snarkos-cli/test_network",
+  "snarkos-node/test_network"
```
Why do we only have to activate these two subdirectories (cli and node)?

Given that we didn't activate `snarkos-node/test_targets` before, was that effectively some dead configuration code?

vicsn replied:
@kpandl this question still stands - is `snarkos-node/test_network` actually used? Should it be called from `cli/Cargo.toml` or not?

kpandl replied:
I don't think `snarkos-node/test_network` is currently used - I've printed the consensus version heights and the coinbase target in `fn check_next_block` of snarkVM. Based on that, in the current state of the codebase (where we neither trigger `snarkos-node/test_network` in the root cargo, nor in the cli cargo), the `test_network` feature works.

vicsn replied:
Should we remove it (and its children) from `Cargo.toml` then?

kpandl replied:
Sure, addressed here: https://github.com/ProvableHQ/snarkOS/pull/3572/commits/491655f9a0657471551f0c971edef554c468093b
Verified it works by printing values in `fn check_next_block` of snarkVM.

kpandl committed:
don't activate snarkos-node/test_network

kpandl committed:
new snarkVM test ref and _heights instead of _versions

kpandl committed:
mainnet bail message

kpandl committed:
update snarkvm ref

vicsn reviewed (changes requested):
Lets wait with merging until snarkVM test_migration_heights is fixed

kpandl committed:
update snarkvm ref

kpandl committed:
Merge branch 'staging' into test-feature

Signed-off-by: Konstantin Pandl <konstantinpandl@yahoo.de>

kpandl committed:
remove test_network from node
